### PR TITLE
ENH: Avoid memory peak and useless computations when printing a MaskedArray.

### PR DESCRIPTION
**Pull request by saimn**
Ref #3544. When printing a `MaskedArray`, the whole array is converted to the object dtype, whereas only a few values are printed to screen. 
The approach here is to cut the array along each axis and keep only a subset that it used for the string conversion. This way the output should not change. The shape used for the cut (100 values for each axis) was chosen so we have enough values when printing on large screen (and the number of printed values depend on the dtype if I understand correctly), maybe there is a better value to choose (inspired from what ndarray.**str** does ?).
Maybe there is a better approach, in which case I am happy to improve this PR.

### CODE REVIEW

**Comment by charris** [comment]
I'm curious about this
```
# convert to object array to make filled work
```
Without looking through the whole file, I'm guessing this is to allow printing `---` for masked values. If we could avoid the object conversion that would be a better solution. Probably more work though...

**Comment by seberg** [comment]
Holy, wow. that is the ugliest thing, but I think that might just solve gh-6723, or maybe it was always used like that?

**Comment by seberg** [comment]
Nvm (I know it probably did not make sense anyway to anyone ;)), but I got it the wrong way around,  ithought the fill value was the problem, but the array is....

**Comment by saimn** [comment]
> Without looking through the whole file, I'm guessing this is to allow printing --- for masked values. If we could avoid the object conversion that would be a better solution. Probably more work though...

Yes, I don't know what are the other possibility to fill the masked values with `--` without doing it manually.
So it is still very ugly but at least it will not eat all the memory (I was always surprised that it was so long to print big arrays).

**Comment by charris** [comment]
There is a back compatibility problem with getting rid of the singleton `masked_print_option` but we could probably do it. I think it should be possible to dispense with it with a bit of cleverness. However, this looks like a good workaround for what we have now. It would be good to use a more descriptive name for `nval` and make it a class variable with an explanation of what it determines. That way when someone complains that it is the wrong value it will be easy to find and change. Could maybe call it `_print_width` or some such.

**Comment by saimn** [comment]
@charris : ok, done. Is it enough with a comment next to the class variable ?

File: numpy/ma/core.py [hunk 1]
```diff
@@ -3710,8 +3712,19 @@ def __str__(self):
...
                 if names is None:
-                    res = self._data.astype("O")
-                    res.view(ndarray)[m] = f
+                    data = self._data
+                    mask = m
+                    # For big arrays, to avoid a costly conversion to the
+                    # object dtype, extract the corners before the conversion.
+                    for axis in range(self.ndim):
+                        if data.shape[axis] > self._print_width:
+                            ind = np.int(self._print_width / 2)
```
Comment (charris) [inline numpy/ma/core.py]: Do `ind = self._print_width // 2` to avoid the integer cast.

Comment (charris) [inline numpy/ma/core.py]: Note that we always use the Python 3 meaning of `/` in order to be able to use the same source code for both Python 2 and 3.

**Comment by charris** [comment]
Comment looks fine, but there is another nit to pick.

**Comment by saimn** [comment]
Good point, thanks !

**Comment by charris** [comment]
Thanks Simon.

**Comment by ahaldane** [comment]
This PR is a great idea! (at least for now)

Possibly for a "real" fix (to avoid to conversion to object type) we need a new dtype system where we can override the `__repr__` function of the dtype. Currently this is only possible for the `np.void` dtype.

**Pull request merged**